_posts/2017-05-30-database-encryption.markdown

```diff
@@ -0,0 +1,172 @@
+     SY1Iv9uaIz8klTXU9YzYtsfUmgXzw7K8StPdbEbo8F1J3JPJB4D7QHF0ObIaSWuf
+     suZqLsvWlYGuJeyX2ntlBN82ORfvUdOrdrbDlmPyj4PfFVl0AK3U3Ai374DNrjKR
+     hF6YFm4TLDaJhUjeV5C43kbE1N2FAMS9LYtPJ44NAoGAFDGHZ/E+aCLerddfwwun
+     MBS6MnftcLPHTZ1RimTrNfsBXipBw1ItWEvn5s0kCm9X24PmdNK4TnhqHYaF4DL5
+     ZjbQK1idEA2Mi8GGPIKJJ2x7P6I0HYiV4qy7fe/w1ZlCXE90B7PuPbtrQY9wO7Ll
+     ipJ45X6I1PnyfOcckn8yafUCgYACtPAlgjJhWZn2v03cTbqA9nHQKyV/zXkyUIXd
+     /XPLrjrP7ouAi5A8WuSChR/yx8ECRgrEM65Be3qBEtoGCB4AS1G0NcigM6qhKBFi
+     VS0aMXr3+V8argcUIwJaWW/x+p2go48yXlJpLHPweeXe8mXEt4iM+QZte6p2yKQ4
+     h9PGQQKBgQCqSydmXBnXGIVTp2sH/2GnpxLYnDBpcJE0tM8bJ42HEQQgRThIChsn
+     PnGA91G9MVikYapgI0VYBHQOTsz8rTIUzsKwXG+TIaK+W84nxH5y6jUkjqwxZmAz
+     r1URaMAun2PfAB4g2N/kEZTExgeOGqXjFhvvjdzl97ux2cTyZhaTXg==
+     -----END RSA PRIVATE KEY-----
+
+  # List Symmetric Key files in the order of current / latest first
+  ciphers:
+     -
+        # Filename containing Symmetric Encryption Key encrypted using the
+        # RSA public key derived from the private key above
+        key_filename: /etc/rails/.rails.key
+        iv_filename:  /etc/rails/.rails.iv
+
+        # Encryption cipher_name
+        #   Recommended values:
+        #      aes-256-cbc
+        #         256 AES CBC Algorithm. Very strong
+        #         Ruby 1.8.7 MRI Approximately 100,000 encryptions or decryptions per second
+        #         JRuby 1.6.7 with Ruby 1.8.7 Approximately 22,000 encryptions or decryptions per second
+        #      aes-128-cbc
+        #         128 AES CBC Algorithm. Less strong.
+        #         Ruby 1.8.7 MRI Approximately 100,000 encryptions or decryptions per second
+        #         JRuby 1.6.7 with Ruby 1.8.7 Approximately 22,000 encryptions or decryptions per second
+        cipher_name:  aes-256-cbc
+
+```
+
+### On Heroku
+Since we don't have access to file system in Heroku outside of source code, `symmetric-encryption` suggests to set encryption key as an environment variable. Configuration is same as above except we replace `key_filename` with `encrypted_key: "<%= ENV['PRODUCTION_ENCRYPTION_KEY1'] %>"`.
+
+You can use following rake task to generate Heroku specific configuration file:
+```
+rails g symmetric_encryption:heroku_config
+```
+This will create `config/symmetric-encryption.yml` file and also output commands you can run to set _encrypted_ encryption key on Heroku as an environment variable.
+
+
+For someone in order to be able to decode the data:
+
+
+- They have to be able to access our database.
+- They need to be able to access our source code.
+- They need to be able to access key files on disk or Heroku's configuration.
+
+### ActiveRecord Integration
+Symmetric Encryption provides a seamless integration with `ActiveRecord`, we can use `attr_encrypted` helper method to define _encrypted_ fields. Lets say we wanted to encrypt a `Note` model that has `note` and `subject`, you can add following to your `ActiveRecord` model:
```

Comment:
...with `ActiveRecord`. We can use...

Let's say... and `subject`. You can add the following...